sn_networking/src/metrics/bad_node.rs

```diff
@@ -77,38 +111,158 @@ impl TimeFrameType {
+        if self.close_group_peers.contains(&peer) {
+            if !self.close_group_peers_that_have_shunned_us.contains(&peer) {
+                self.metric_current_group.inc();
+                self.close_group_peers_that_have_shunned_us.insert(peer);
+            }
+        } else if self.old_close_group_peers.iter().any(|(p, _)| p == &peer)
+            && !self
+                .old_close_group_peers_that_have_shunned_us
+                .contains(&peer)
+        {
+            self.metric_old_group.inc();
+            self.old_close_group_peers_that_have_shunned_us.insert(peer);
+        }
+    }
+
+    pub(crate) fn update_close_group_peers(&mut self, new_closest_peers: Vec<PeerId>) {
+        let new_members: Vec<PeerId> = new_closest_peers
+            .iter()
+            .filter(|p| !self.close_group_peers.contains(p))
+            .cloned()
+            .collect();
+        let evicted_members: Vec<PeerId> = self
+            .close_group_peers
+            .iter()
+            .filter(|p| !new_closest_peers.contains(p))
+            .cloned()
+            .collect();
+        for new_member in &new_members {
+            // if it has shunned us before, update the metrics.
+            if self
+                .old_close_group_peers_that_have_shunned_us
+                .contains(new_member)
+            {
+                self.metric_old_group.dec();
+                self.old_close_group_peers_that_have_shunned_us
+                    .remove(new_member);
+
+                self.metric_current_group.inc();
+                self.close_group_peers_that_have_shunned_us
+                    .insert(*new_member);
+            }
+        }
+
+        for evicted_member in &evicted_members {
+            self.old_close_group_peers
+                .push((*evicted_member, Instant::now()));
+
+            // if it has shunned us before, update the metrics.
+            if self
+                .close_group_peers_that_have_shunned_us
+                .contains(evicted_member)
+            {
+                self.metric_current_group.dec();
+                self.close_group_peers_that_have_shunned_us
+                    .remove(evicted_member);
+
+                self.metric_old_group.inc();
+                self.old_close_group_peers_that_have_shunned_us
+                    .insert(*evicted_member);
+            }
+        }
 
-    pub fn record_shunned_metric(&mut self) {
+        if !new_members.is_empty() {
+            debug!("The close group has been updated. The new members are {new_members:?}. The evicted members are {evicted_members:?}");
+            self.close_group_peers = new_closest_peers;
+
+            if self.old_close_group_peers.len() > MAX_EVICTED_CLOSE_GROUP_PEERS {
+                // clean the oldest Instant ones
+                self.old_close_group_peers
+                    .sort_by_key(|(_, instant)| std::cmp::Reverse(*instant));
```

Comment:
a VecDeque then call `push_back` and `pop_front` shall achieve the same?
then don't need the timestamp ?

Reply:
Aye, that will work! Thanks